src/lib/xmlparse.rs

```diff
@@ -833,7 +833,7 @@ pub struct XML_ParserStruct<'scf> {
     pub m_declAttributeId: *mut ATTRIBUTE_ID,
     pub m_declAttributeIsCdata: XML_Bool,
     pub m_declAttributeIsId: XML_Bool,
-    pub m_dtd: *mut DTD<'scf>,
+    pub m_dtd: Rc<DTD>,
```

Comment:
I take it `Rc<RefCell<DTD>>` wasn't viable here to avoid the individual field Cells and RefCells in DTD?

Reply:
A lot of the code just accesses one or a few of the `Cell`s at a time without touching the `RefCell`s, so this split is actually zero-cost.

Reply:
There's actually a whole discussion on the granularity of the split here, with 3 main options imho (from coarsest to finest-grained):

1. Put all of `DTD` in a single `RefCell`. I really don't like this one because that structure has a lot of scalar fields that get read individually in a lot of places, e.g., `standalone`, and it would add extra overhead to `borrow_mut` for each one of those accesses. Plus, recursive parser calls would run into cases where the inner parser would try to re-borrow an already borrowed `DTD`.

2. The current solution, which works.

3. Wrap every single field of `DTD` in its own `Cell` or `RefCell`. I tried to group them instead so there wouldn't be so many borrows, since they're slow. I still had to make sure that those borrows aren't held across potentially recursive calls.

Reply:
I think what you've got right now makes sense. It's a little weird at first glance so maybe a comment explaining why like you did above would be nice?

Reply:
Sure, I can add a detailed comment.